### PR TITLE
Rename permitrbf to mempoolreplacement and provide minimal string-list forward compatibility (needs 0.12 backport)

### PR DESCRIPTION
**Pull request by luke-jr**
The purpose of supporting -replacebyfee=opt-in is so that future versions can also support things like -replacebyfee=fss or -replacebyfee=opt-in,fss (for First-Seen-Safe behaviour)

### CODE REVIEW

**Comment by jonasschnelli** [comment]
The forward compatibility looks good to me. It's a tiny bit speculative (I'm not aware of plans to change  [distinct between multiple]) RBF options.

utACK

**Comment by laanwj** [comment]
> (I'm not aware of plans to change [distinct between multiple]) RBF options.

There are no such plans, but @luke-jr wants to support other kinds of RBF in his own fork.
No strong opinion either for or against this.

**Comment by MarcoFalke** [comment]
How would this work? The release notes mention `-permitrbf` and you removed that code entirely.

**Comment by jonasschnelli** [comment]
@MarcoFalke: Right. The release notes need an update – IF – we would merge this.
IIUTC: The forward compatibility means, that of someone throws in string arg like `-replacebyfee=foo,bar,opt-in` it would result in enabling opt-in-RBF.

File: src/init.cpp [hunk 1]
```diff
@@ -1030,7 +1032,20 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
     nMaxTipAge = GetArg("-maxtipage", DEFAULT_MAX_TIP_AGE);
-    fPermitReplacement = GetBoolArg("-permitrbf", DEFAULT_PERMIT_REPLACEMENT);
+
+    fEnableReplacement = GetBoolArg("-replacebyfee", DEFAULT_ENABLE_REPLACEMENT);
+    if ((!fEnableReplacement) && mapArgs.count("-replacebyfee")) {
+        // Minimal effort at forwards compatibility
+        std::string strReplacementModeList = GetArg("-replacebyfee", "");  // default is impossible
+        std::vector<std::string> vstrReplacementModes;
+        boost::split(vstrReplacementModes, strReplacementModeList, boost::is_any_of(","));
+        BOOST_FOREACH(const std::string& strReplacementMode, vstrReplacementModes) {
```
Comment (jonasschnelli) [inline src/init.cpp]: `fEnableReplacement = (std::find(vstrReplacementModes.begin(), vstrReplacementModes.end(), "opt-in") != v.end())` (and avoid the foreach)?

Comment (laanwj) [inline src/init.cpp]: Yes that should be the same, as this is in `if(!fEnableReplacement)` so it can never be set to false by this.

**Comment by MarcoFalke** [comment]
> – IF – we would merge this.

It should be part of this pull; Otherwise we'd end up in an inconsistent state. Also adding it to the blacklist may be worth to consider as the option is already in use and promoted/posted.

**Comment by luke-jr** [comment]
@jonasschnelli RBF-FSS at present seems more popular than opt-in, so I would not be surprised if someone implemented it for 0.13.

@MarcoFalke This is the PR against master, so it cannot include changes to 0.12 release notes.

**Comment by luke-jr** [comment]
Added @jonasschnelli 's suggestion.

**Comment by laanwj** [comment]
Suggestion by @gmaxwell on IRC: Though if it's pluggable policy for luke I wonder why he didn't call it -mempoolreplace 

I think that makes sense. If this selects a "mempool replacement policy", by fee is just one option.

**Comment by luke-jr** [comment]
Good idea, done.

File: src/init.cpp [hunk 2]
```diff
@@ -1030,7 +1032,15 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
...
     nMaxTipAge = GetArg("-maxtipage", DEFAULT_MAX_TIP_AGE);
-    fPermitReplacement = GetBoolArg("-permitrbf", DEFAULT_PERMIT_REPLACEMENT);
+
+    fEnableReplacement = GetBoolArg("-mempoolreplacement", DEFAULT_ENABLE_REPLACEMENT);
+    if ((!fEnableReplacement) && mapArgs.count("-mempoolreplacement")) {
+        // Minimal effort at forwards compatibility
+        std::string strReplacementModeList = GetArg("-mempoolreplacement", "");  // default is impossible
+        std::vector<std::string> vstrReplacementModes;
+        boost::split(vstrReplacementModes, strReplacementModeList, boost::is_any_of(","));
+        fEnableReplacement = (std::find(vstrReplacementModes.begin(), vstrReplacementModes.end(), "fee") != vstrReplacementModes.end());
```
Comment (laanwj) [inline src/init.cpp]: I'm not sure 'fee' is a good name. How would you then call non-opt in RBF?

**Pull request merged**